### PR TITLE
Allow setting config file name via CLI

### LINKED ISSUE #208: --config=path/to/config for a custom CPPLINT.cfg
I'd like to be able to put he cpplint(1) config file in a directory that is specific for configuration files in the repo.  Having it in the root of the repository unnecessarily clutters the directory in a ls(1).

Would you mind adding a command line flag for passing a custom location for the file?

I'd like to be able to use the following recipe in my makefiles:

```
	cpplint --config=./etc/cpplint/cpplint.cfg $< >/dev/null
```

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
This introduces a new command-line option that allows configuration overrides to be read from a file with a name other than `CPPLINT.cfg`.  This can be useful if one wants to have separate "normal" and "noisy" modes, for example:

```console
  cpplint --config=.cpplint <file>
  cpplint --config=.cpplint-noisy <file>
```

It's also nice if you'd prefer a config file named `.cpplint` to a non-dotfile named in ALL CAPS.

---

I have tested this manually, but I haven't looked into adding testcases yet, as that's a bigger job than these changes were.  I'm willing to do that work, but my time is somewhat limited.

### CODE REVIEW

**Comment by tkruse** [comment]
Hi, thanks for the contribution. Can you explain a bit more why a project would want two different configuration for the same files? When would you want to use a more "noisy" configuration?

Regarding the dotfile name instead of CPPLINT.cfg, let's create a separate ticket.

The nice thing about a fixed filename in theory is that other toolchains could pick it up, having custom names confuses other tools. Not that I am aware of any such integration with cpplint.

For usability, I think it might be good to verify that the value passed for `--config` is a basename without a folder (if that's most useful). Also if users specify a filename, should it be an error if the file is not found ever?

I believe there are currently no automated test for CPPLINT.cfg. If you can just add sample files into `/samples` with a noisy and a less noisy cfg file, I can write a regression test. The source files should produce some expected cpplint errors, of course.

**Comment by gedankenexperimenter** [comment]
> Can you explain a bit more why a project would want two different configuration for the same files? When would you want to use a more "noisy" configuration?

The idea is that we would use a "standard" config for automated checking in CI (e.g. GitHub workflows), and a stricter config to be run manually when looking for areas that can be improved.  Because the config file overrides the command-line filters, in order to run that stricter test, it would be necessary to first delete or rename `CPPLINT.cfg` whenever running cpplint with the alternate config.  If command-line options took precedence over the config file, as they do for most utilities, this wouldn't be as helpful, but that would be a much more extensive design change for cpplint.

> The nice thing about a fixed filename in theory is that other toolchains could pick it up, having custom names confuses other tools. Not that I am aware of any such integration with cpplint.

With this change, the default config filename is still `CPPLINT.cfg`; I would expect anyone who goes out of their way to change it could handle repercussions of that change.

> For usability, I think it might be good to verify that the value passed for `--config` is a basename without a folder (if that's most useful). Also if users specify a filename, should it be an error if the file is not found ever?

Agreed.  I'll update this PR accordingly.  I don't think it should be an error if no config file is found (any more than it should be an error if `CPPLINT.cfg` isn't found currently), but I don't feel strongly about that.  It would add some complexity to check for it, especially if it is decided that such an error (or warning) should only be produced if the config file name is changed from the default.

**Comment by tkruse** [comment]
So it seems to me that other tools would solve that goal by having different priority for output, like "error" vs. "warning". Warnings would not fail the check, but generate additional output.

It would seem painful to me to try to keep two different files in sync as well. I don't know how much effort it would be to add different error priorities to cpplint, but it would seem to me that that would be a better way to solve the requirement.

**Comment by gedankenexperimenter** [comment]
I'm unfamiliar with the testing requirements for tox for this repo, and it's not really worth my time to familiarize myself with it, as I'm unlikely to continue making future contributions here.  If someone else wants to do that work so this PR can be merged, that sounds great.  Otherwise, I will just continue to use my fork for the project that I need this change for.

File: cpplint.py [hunk 1]
```diff
@@ -6781,6 +6789,11 @@ def ParseArguments(args):
       recursive = True
     elif opt == '--includeorder':
       ProcessIncludeOrderOption(val)
+    elif opt == '--config':
+      global _config_filename
+      _config_filename = val
+      if os.path.basename(_config_filename) != _config_filename:
+        PrintUsage('Config file name must not include directory components.')
```
Comment (jayvdb) [inline cpplint.py:6796]: IMO this isnt a good restriction, but I will merge as-is and @ACSimon33 maybe you want to remove it and possibly add some tests about this feature in a followup PR?

Comment (nlohmann) [reply cpplint.py:6796]: If I understand this correctly, then removing this check would fix #208, right?

**Pull request merged**